File: python/src/smooai_config/config_manager.py
"""Unified configuration manager merging file, remote, and env sources."""

import os
import sys
import threading
import time
from typing import Any

from smooai_config.client import ConfigClient
from smooai_config.env_config import find_and_process_env_config
from smooai_config.file_config import find_and_process_file_config
from smooai_config.merge import merge_replace_arrays


class ConfigManager:
    """Unified config manager that merges three sources in precedence order.

    Merge precedence (highest to lowest):
    1. Env vars -- always win
    2. Remote API -- authoritative env-specific values from server
    3. File config -- base defaults from JSON files

    Thread-safe. Lazy initialization loads and merges all sources on first access.
    Per-key caches with configurable TTL for each tier (public, secret, feature_flag).
    """

    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
        org_id: str | None = None,
        environment: str | None = None,
        schema_keys: set[str] | None = None,
        env_prefix: str = "",
        schema_types: dict[str, str] | None = None,
        cache_ttl: float = 86400,  # 24 hours
        env: dict[str, str] | None = None,
    ) -> None:
        self._lock = threading.RLock()
        self._initialized = False
        self._config: dict[str, Any] = {}
        self._public_cache: dict[str, tuple[Any, float]] = {}
        self._secret_cache: dict[str, tuple[Any, float]] = {}
        self._feature_flag_cache: dict[str, tuple[Any, float]] = {}
        self._api_key = api_key
        self._base_url = base_url
        self._org_id = org_id
        self._environment = environment
        self._schema_keys = schema_keys
        self._env_prefix = env_prefix
        self._schema_types = schema_types
        self._cache_ttl = cache_ttl
        self._env = env

    def _resolve_env_var(self, key: str) -> str | None:
        """Resolve an environment variable from the configured env dict or os.environ."""
        if self._env is not None:
            return self._env.get(key)
        return os.environ.get(key)

    def _initialize(self) -> None:
        """Lazy init: load file config, remote config, and env config, then merge."""
        if self._initialized:
            return

        # 1. Load file config
        file_config = find_and_process_file_config(
            schema_keys=self._schema_keys,
            env=self._env,
        )

        # 2. Load env config
        env_config = find_and_process_env_config(
            schema_keys=self._schema_keys or set(),
            prefix=self._env_prefix,
            schema_types=self._schema_types,
            env=self._env,
        )

        # 3. Try remote fetch if API creds available
        remote_config: dict[str, Any] = {}
        api_key = self._api_key or self._resolve_env_var("SMOOAI_CONFIG_API_KEY")
        base_url = self._base_url or self._resolve_env_var("SMOOAI_CONFIG_API_URL")
        org_id = self._org_id or self._resolve_env_var("SMOOAI_CONFIG_ORG_ID")

        if api_key and base_url and org_id:
            # Resolve environment: explicit param > env var > default "development"
            resolved_environment = self._environment or self._resolve_env_var("SMOOAI_CONFIG_ENV") or "development"

            try:
                client = ConfigClient(
                    base_url=base_url,
                    api_key=api_key,
                    org_id=org_id,
                    environment=resolved_environment,
                )
                try:
                    remote_config = client.get_all_values()
                finally:
                    client.close()
            except Exception as exc:
                print(
                    f"[Smooai Config] Warning: Failed to fetch remote config: {exc}",
                    file=sys.stderr,
                )

        # 4. Merge: file < remote < env
        merged: dict[str, Any] = merge_replace_arrays({}, file_config)
        merged = merge_replace_arrays(merged, remote_config)
        merged = merge_replace_arrays(merged, env_config)
        self._config = merged
        self._initialized = True

    def _get_from_cache(self, cache: dict[str, tuple[Any, float]], key: str) -> tuple[bool, Any]:
        """Get value from TTL cache. Returns (hit, value)."""
        if key in cache:
            value, expires_at = cache[key]
            if time.time() < expires_at:
                return True, value
            del cache[key]
        return False, None

    def _set_cache(self, cache: dict[str, tuple[Any, float]], key: str, value: Any) -> None:
        """Set value in TTL cache."""
        cache[key] = (value, time.time() + self._cache_ttl)

    def _get_value(self, key: str, cache: dict[str, tuple[Any, float]]) -> Any | None:
        """Get config value from merged config with per-tier caching."""
        with self._lock:
            hit, value = self._get_from_cache(cache, key)
            if hit:
                return value

            self._initialize()

            value = self._config.get(key)
            self._set_cache(cache, key, value)
            return value

    def get_public_config(self, key: str) -> Any | None:
        """Retrieve a public config value."""
        return self._get_value(key, self._public_cache)

    def get_secret_config(self, key: str) -> Any | None:
        """Retrieve a secret config value."""
        return self._get_value(key, self._secret_cache)

    def get_feature_flag(self, key: str) -> Any | None:
        """Retrieve a feature flag value."""
        return self._get_value(key, self._feature_flag_cache)

    def invalidate(self) -> None:
        """Clear all caches and force re-initialization on next access."""
        with self._lock:
            self._initialized = False
            self._config = {}
            self._public_cache.clear()
            self._secret_cache.clear()
            self._feature_flag_cache.clear()

File: go/config/config_manager.go
package config

import (
	"fmt"
	"os"
	"sync"
	"time"
)

// ConfigManager is a unified config manager that merges three sources in this
// precedence (highest to lowest):
//
//  1. Env vars — always win
//  2. Remote API — authoritative values from server
//  3. File config — base defaults from JSON files
//
// Thread-safe via sync.Mutex. Lazy initialization loads all sources on first access.
// Per-key caches with configurable TTL for each tier (public, secret, feature_flag).
type ConfigManager struct {
	mu          sync.Mutex
	initialized bool
	config      map[string]any // single merged config

	// Per-tier caches
	publicCache map[string]localCacheEntry
	secretCache map[string]localCacheEntry
	ffCache     map[string]localCacheEntry

	// Local config params
	schemaKeys  map[string]bool
	envPrefix   string
	schemaTypes map[string]string
	cacheTTL    time.Duration
	envOverride map[string]string // for testing

	// Remote API params
	apiKey      string
	baseURL     string
	orgID       string
	environment string
}

// ConfigManagerOption is a functional option for ConfigManager.
type ConfigManagerOption func(*ConfigManager)

// NewConfigManager creates a new unified config manager with functional options.
func NewConfigManager(opts ...ConfigManagerOption) *ConfigManager {
	m := &ConfigManager{
		publicCache: make(map[string]localCacheEntry),
		secretCache: make(map[string]localCacheEntry),
		ffCache:     make(map[string]localCacheEntry),
		cacheTTL:    defaultLocalCacheTTL,
	}
	for _, opt := range opts {
		opt(m)
	}
	return m
}

// WithAPIKey sets the API key for remote config fetching.
func WithAPIKey(key string) ConfigManagerOption {
	return func(m *ConfigManager) { m.apiKey = key }
}

// WithBaseURL sets the base URL for the remote config API.
func WithBaseURL(url string) ConfigManagerOption {
	return func(m *ConfigManager) { m.baseURL = url }
}

// WithOrgID sets the organization ID for remote config fetching.
func WithOrgID(id string) ConfigManagerOption {
	return func(m *ConfigManager) { m.orgID = id }
}

// WithConfigEnvironment sets the environment name for remote config fetching.
func WithConfigEnvironment(env string) ConfigManagerOption {
	return func(m *ConfigManager) { m.environment = env }
}

// WithCMSchemaKeys sets schema keys for env config filtering.
func WithCMSchemaKeys(keys map[string]bool) ConfigManagerOption {
	return func(m *ConfigManager) { m.schemaKeys = keys }
}

// WithCMEnvPrefix sets env var prefix for stripping.
func WithCMEnvPrefix(prefix string) ConfigManagerOption {
	return func(m *ConfigManager) { m.envPrefix = prefix }
}

// WithCMSchemaTypes sets schema type hints for coercion.
func WithCMSchemaTypes(types map[string]string) ConfigManagerOption {
	return func(m *ConfigManager) { m.schemaTypes = types }
}

// WithCMCacheTTL sets the cache TTL.
func WithCMCacheTTL(ttl time.Duration) ConfigManagerOption {
	return func(m *ConfigManager) { m.cacheTTL = ttl }
}

// WithCMEnvOverride overrides environment variables (for testing).
func WithCMEnvOverride(env map[string]string) ConfigManagerOption {
	return func(m *ConfigManager) { m.envOverride = env }
}

// getEnvVal looks up a key from the env override map, falling back to os.Getenv.
func (m *ConfigManager) getEnvVal(key string) string {
	if m.envOverride != nil {
		return m.envOverride[key]
	}
	return os.Getenv(key)
}

func (m *ConfigManager) initialize() error {
	if m.initialized {
		return nil
	}

	// Resolve the env map for file/env config functions
	var env map[string]string
	if m.envOverride != nil {
		env = m.envOverride
	} else {
		env = osEnvMap()
	}

	// 1. Load file config (graceful — file config is optional)
	fileConfig, err := findAndProcessFileConfigWithEnv(env)
	if err != nil {
		fileConfig = make(map[string]any)
	}

	// 2. Load env config
	schemaKeys := m.schemaKeys
	if schemaKeys == nil {
		schemaKeys = make(map[string]bool)
	}
	envConfig := findAndProcessEnvConfigWithEnv(schemaKeys, m.envPrefix, m.schemaTypes, env)

	// 3. Try remote fetch if API creds are available
	remoteConfig := make(map[string]any)

	apiKey := m.apiKey
	baseURL := m.baseURL
	orgID := m.orgID

	// Check env vars as fallback for API credentials
	if apiKey == "" {
		apiKey = m.getEnvVal("SMOOAI_CONFIG_API_KEY")
	}
	if baseURL == "" {
		baseURL = m.getEnvVal("SMOOAI_CONFIG_API_URL")
	}
	if orgID == "" {
		orgID = m.getEnvVal("SMOOAI_CONFIG_ORG_ID")
	}

	if apiKey != "" && baseURL != "" && orgID != "" {
		// Resolve environment
		configEnv := m.environment
		if configEnv == "" {
			configEnv = m.getEnvVal("SMOOAI_CONFIG_ENV")
		}
		if configEnv == "" {
			configEnv = "development"
		}

		client := NewConfigClient(baseURL, apiKey, orgID)
		defer client.Close()

		values, err := client.GetAllValues(configEnv)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[Smooai Config] Warning: Failed to fetch remote config: %v\n", err)
		} else {
			remoteConfig = values
		}
	}

	// 4. Merge: file < remote < env
	merged := MergeReplaceArrays(make(map[string]any), fileConfig).(map[string]any)
	merged = MergeReplaceArrays(merged, remoteConfig).(map[string]any)
	merged = MergeReplaceArrays(merged, envConfig).(map[string]any)

	m.config = merged
	m.initialized = true
	return nil
}

func (m *ConfigManager) getFromTier(key string, cache map[string]localCacheEntry) (any, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	// Check cache
	if entry, ok := cache[key]; ok {
		if time.Now().Before(entry.expiresAt) {
			return entry.value, nil
		}
		delete(cache, key)
	}

	// Initialize if needed
	if err := m.initialize(); err != nil {
		return nil, err
	}

	// Lookup in merged config
	value := m.config[key]

	// Cache the result
	cache[key] = localCacheEntry{value: value, expiresAt: time.Now().Add(m.cacheTTL)}
	return value, nil
}

// GetPublicConfig retrieves a public config value.
func (m *ConfigManager) GetPublicConfig(key string) (any, error) {
	return m.getFromTier(key, m.publicCache)
}

// GetSecretConfig retrieves a secret config value.
func (m *ConfigManager) GetSecretConfig(key string) (any, error) {
	return m.getFromTier(key, m.secretCache)
}

// GetFeatureFlag retrieves a feature flag value.
func (m *ConfigManager) GetFeatureFlag(key string) (any, error) {
	return m.getFromTier(key, m.ffCache)
}

// Invalidate clears all caches and forces re-initialization on next access.
func (m *ConfigManager) Invalidate() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.initialized = false
	m.config = nil
	m.publicCache = make(map[string]localCacheEntry)
	m.secretCache = make(map[string]localCacheEntry)
	m.ffCache = make(map[string]localCacheEntry)
}
